src/xtgeo/xyz/_xyz_roxapi.py

```diff
@@ -305,237 +399,139 @@ def _roxapi_export_xyz_viafile(
+def _get_attribute_names_from_dataframe(df: pd.DataFrame) -> list[str]:
+    return [col for col in df.columns if col not in XYZ_COLUMNS]
 
-    In some case, attribute columns get dtype 'object' while they clearly
-    represents a numerical property (int or float). Here the pandas to_numerics()
-    function is applied per attribute column, and will do a conversion if possible;
-    otherwise the 'object' dtype will be preserved.
-    """
-    if len(dfr) < 1:
-        return dfr
 
-    newdfr = dfr.copy()
-    for name in dfr.columns[3:]:
-        newdfr[name] = pd.to_numeric(dfr[name], errors="ignore")
-    return newdfr
-
-
-def _check_category_etc(
-    proj,
-    name,
-    category,
-    stype,
-    realisation,
-    mode="get",
-):
-    """Helper to check if valid placeholder' whithin RMS."""
+def _get_attribute_type_from_values(values: np.ndarray) -> str:
+    if "float" in str(values.dtype):
+        return "float"
+    if "int" in str(values.dtype):
+        return "int"
+    return "str"
 
-    logger.warning("Realisation %s not in use", realisation)
 
-    stypedict = {"horizons": proj.horizons, "zones": proj.zones, "faults": proj.faults}
-
-    if stype in stypedict:
-        if name not in stypedict[stype]:
-            logger.error("Cannot access name in stype=%s: %s", stype, name)
-            return False
-        if category not in stypedict[stype].representations:
-            logger.error("Cannot access category in stype=%s: %s", stype, category)
-            return False
-
-    elif stype in ("clipboard", "general2d_data") and mode == "get":
-        folders = None
-        if category:
-            if isinstance(category, list):
-                folders = category
-            elif isinstance(category, str) and "|" in category:
-                folders = category.split("|")
-            elif isinstance(category, str) and "/" in category:
-                folders = category.split("/")
-            elif isinstance(category, str):
-                folders = []
-                folders.append(category)
-            else:
-                raise RuntimeError(
-                    f"Cannot parse category: {category}, see documentation!"
-                )
-            try:
-                roxxyz = getattr(proj, stype).folders[folders]
-            except KeyError as keyerr:
-                logger.error(
-                    "Cannot access clipboards folder (not existing?): %s", keyerr
-                )
-                return False
-        else:
-            roxxyz = proj.clipboard
+def _replace_undefined_values(values: np.ndarray, dtype: str = None) -> np.ndarray:
+    """Set xtgeo UNDEF values to np.nan or empty string dependent on type"""
+    values = pd.to_numeric(values, errors="ignore")
 
-        if name not in roxxyz:
-            raise ValueError(f"Name {name} is not within Clipboard...")
+    dtype = dtype if dtype is not None else _get_attribute_type_from_values(values)
```

Comment:
```suggestion
    dtype = dtype or _get_attribute_type_from_values(values)
```

Reply:
I've never gotten used to this 'or' syntax! But looks much more elegant 🙂